synapse/rest/media/upload_resource.py

```diff
@@ -103,3 +114,53 @@ async def on_POST(self, request: SynapseRequest) -> None:
         respond_with_json(
             request, 200, {"content_uri": str(content_uri)}, send_cors=True
         )
+
+
+class AsyncUploadServlet(BaseUploadServlet):
+    PATTERNS = [
+        re.compile(
+            "/_matrix/media/v1/upload/(?P<server_name>[^/]*)/(?P<media_id>[^/]*)$"
```

Comment:
I was surprised to see that this is a [v3 endpoint according to the spec](https://spec.matrix.org/v1.8/client-server-api/#put_matrixmediav3uploadservernamemediaid)

```suggestion
            "/_matrix/media/v3/upload/(?P<server_name>[^/]*)/(?P<media_id>[^/]*)$"
```

I think it is worth keeping different patterns so that r0/v1 don't work though.

Reply:
Yeah, the versioning scheme is stupid. Create is `/v1` while this is `/v3` presumably because `/v1/upload` already existed and could conflict? We should have just made both of them `/v3` so it's not confusing, but Travis disagrees and that's what ended up in the spec.